### PR TITLE
Fix/ handle long pvc names

### LINKED ISSUE #155: Handle long named PVs better
kube-prometheus-stack Helm chart produces some stupid naming conventions, that fly very close to the sun regarding name length. Running `pvmigrate` throws an error as the new PVC name goes over the max length.

It would be a nice to have for pvmigrate to handle these situations and maybe derive a shorter name for the migration PVC

```
Running pvmigrate build:
version=v0.8.0-dirty
sha=33bbb070fd5685c16b544c84f7ec8e7398c8164d
time=2023-02-25T11:48:51Z

...

Found 1 matching PVCs to migrate across 1 namespaces:
namespace: pvc:                                                                                         pv:                                      size:
prometheus prometheus-kube-prometheus-stack-prometheus-db-prometheus-kube-prometheus-stack-prometheus-0 pvc-24dd3b63-f891-413c-9a33-3acaefe5c0e0 20Gi
prometheus kube-prometheus-stack-grafana                                                                pvc-aa1153a7-dc83-445e-95fa-4eaf87eeffd4 250Mi
prometheus prometheus-kube-prometheus-stack-prometheus-db-prometheus-kube-prometheus-stack-prometheus-1 pvc-fcf77baf-7c4f-4be9-916f-0206dd7e0392 500Gi

Creating new PVCs to migrate data to using the nfs-client-standard StorageClass
migration failed: failed to create new PVC prometheus-kube-prometheus-stacus-stack-prometheus-0-pvcmigrate in prometheus: PersistentVolumeClaim "prometheus-kube-prometheus-stacus-stack-prometheus-0-pvcmigrate" is invalid: metadata.labels: Invalid value: "prometheus-kube-prometheus-stack-prometheus-db-prometheus-kube-prometheus-stack-prometheus-0": must be no more than 63 characters
```

### PR DESCRIPTION
**Pull request by clementnuss**
- instead of putting the original PVC name into a label, we write it to an
  annotation
- also, PVC name are no longer truncated at 63 chars, as they can be up to 253
  long (DNS name)
- finally, instead the migration's pod name is prefixed with `migrate-` and
  truncated to a max total length of 63 chars (as pod name _should_ typically
  not be longer than 63 chars)

closes #155 #171

### CODE REVIEW

**Comment by clementnuss** [comment]
also, I noticed that the `baseAnnotation=sourcePvcName` isn't actually used when swapping off the PVCs after migration.
please let me know if I missed the usage for this annotation, but I think we can change it from a label to an annotation without trouble.

also, here's a log of a migration for a PVC with a 253 chars long name:
<details>
<summary>pvmigrate_log.txt</summary>

```text
./pvmigrate -source-sc nfs -dest-sc nfs -skip-preflight-validation -namespace=cjmnuss
Running pvmigrate build:
version=(devel)
sha=
time=

Found 1 StorageClass:
nfs

Migrating data from nfs to nfs
PV pvc-641fbe2b-3e2a-49ba-a58b-06090f6e1c3b does not match source SC nfs, not migrating

Found 1 matching PVCs to migrate across 1 namespaces:
namespace: pvc:                                                                                                                                                                                                                                                          pv:                                      size:
cjmnuss    super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonnglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong-pvcmigrate pvc-2e452efe-9733-4f56-8861-9fecc9f9038d 1Gi
cjmnuss    super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong pvc-ead5ef1d-5b07-4152-8e63-b3469f8ab042 1Gi

Creating new PVCs to migrate data to using the nfs StorageClass
created new PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglononglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong-pvcmigrate-pvcmigrate with size 1Gi in cjmnuss
found existing PVC with name super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonnglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong-pvcmigrate, not creating new one

Found 1 matching pods to migrate across 1 namespaces:
namespace: pod:
cjmnuss    test-deploy-attached-to-super-long-pvc-with-a-name-63-charv547h

Scaling down StatefulSets and Deployments with matching PVCs
scaling Deployment test-deploy-attached-to-super-long-pvc-with-a-name-63-char-long-659898f79f from 1 to 0 in cjmnuss

Waiting for pods with mounted PVCs to be cleaned up
Found pod test-deploy-attached-to-super-long-pvc-with-a-name-63-charv547h in cjmnuss mounting to-be-migrated PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong, waiting
All pods removed successfully

Copying data from nfs PVCs to nfs PVCs
Copying data from super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonnglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong-pvcmigrate (pvc-2e452efe-9733-4f56-8861-9fecc9f9038d) to super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglononglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong-pvcmigrate-pvcmigrate in cjmnuss
Determining the node to migrate PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonnglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong-pvcmigrate on
Creating pvc migrator pod on node
waiting for pod migrate--super-long-pvc-253-chaglonglonglonglonglong-pvcmigrate to start in cjmnuss
migrating PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonnglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong-pvcmigrate:
.done!
finished migrating PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonnglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong-pvcmigrate
Copying data from super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong (pvc-ead5ef1d-5b07-4152-8e63-b3469f8ab042) to super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonnglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong-pvcmigrate in cjmnuss
Determining the node to migrate PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong on
Creating pvc migrator pod on node abcd.domain.ch
waiting for pod migrate--super-long-pvc-253-chalonglonglonglonglonglonglonglong to start in cjmnuss
migrating PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong:
..done!
finished migrating PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong

Swapping PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonnglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong-pvcmigrate in cjmnuss to the new StorageClass
Marking original PV pvc-2e452efe-9733-4f56-8861-9fecc9f9038d as to-be-retained
Marking migrated-to PV pvc-3e2be7e1-fa9b-4d32-90f5-5a553a5104bb as to-be-retained
Deleting original PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonnglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong-pvcmigrate in cjmnuss to free up the name
Deleting migrated-to PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglononglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong-pvcmigrate-pvcmigrate in cjmnuss to release the PV
Waiting for original PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonnglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong-pvcmigrate in cjmnuss to finish deleting
Waiting for migrated-to PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglononglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong-pvcmigrate-pvcmigrate in cjmnuss to finish deleting
Removing claimref from original PV pvc-2e452efe-9733-4f56-8861-9fecc9f9038d
Removing claimref from migrated-to PV pvc-3e2be7e1-fa9b-4d32-90f5-5a553a5104bb
Creating new PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonnglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong-pvcmigrate with migrated-to PV pvc-3e2be7e1-fa9b-4d32-90f5-5a553a5104bb
Resetting migrated-to PV pvc-3e2be7e1-fa9b-4d32-90f5-5a553a5104bb reclaim policy
Got conflict updating PV pvc-3e2be7e1-fa9b-4d32-90f5-5a553a5104bb, waiting 5s to retry
Deleting original, now redundant, PV pvc-2e452efe-9733-4f56-8861-9fecc9f9038d
Successfully migrated PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonnglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong-pvcmigrate in cjmnuss from PV pvc-2e452efe-9733-4f56-8861-9fecc9f9038d to pvc-3e2be7e1-fa9b-4d32-90f5-5a553a5104bb

Swapping PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong in cjmnuss to the new StorageClass
Marking original PV pvc-ead5ef1d-5b07-4152-8e63-b3469f8ab042 as to-be-retained
Marking migrated-to PV pvc-3e2be7e1-fa9b-4d32-90f5-5a553a5104bb as to-be-retained
Deleting original PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong in cjmnuss to free up the name
Deleting migrated-to PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonnglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong-pvcmigrate in cjmnuss to release the PV
Waiting for original PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong in cjmnuss to finish deleting
Waiting for migrated-to PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonnglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong-pvcmigrate in cjmnuss to finish deleting
Removing claimref from original PV pvc-ead5ef1d-5b07-4152-8e63-b3469f8ab042
Removing claimref from migrated-to PV pvc-3e2be7e1-fa9b-4d32-90f5-5a553a5104bb
Creating new PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong with migrated-to PV pvc-3e2be7e1-fa9b-4d32-90f5-5a553a5104bb
Resetting migrated-to PV pvc-3e2be7e1-fa9b-4d32-90f5-5a553a5104bb reclaim policy
Deleting original, now redundant, PV pvc-ead5ef1d-5b07-4152-8e63-b3469f8ab042
Successfully migrated PVC super-long-pvc-253-chars-longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong in cjmnuss from PV pvc-ead5ef1d-5b07-4152-8e63-b3469f8ab042 to pvc-3e2be7e1-fa9b-4d32-90f5-5a553a5104bb

Scaling back StatefulSets and Deployments with matching PVCs
scaling Deployment test-deploy-attached-to-super-long-pvc-with-a-name-63-char-long from 0 to 1 in cjmnuss

Success!
```
</details>

**Comment by laverya** [comment]
I'm sorry for leaving this open for so long! I'm taking a look now

File: pkg/k8sutil/truncate.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package k8sutil
+
+import "fmt"
+
+const nameSuffix = "-pvcmigrate"
+
+// if the length after adding the suffix is more than 63 characters, we need to reduce that to fit within k8s limits
```
Comment (laverya) [inline pkg/k8sutil/truncate.go:7]:
```suggestion
// if the length after adding the suffix is more than 253 characters, we need to reduce that to fit within k8s limits
```

File: pkg/k8sutil/truncate.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+// pruning from the end runs the risk of dropping the '0'/'1'/etc of a statefulset's PVC name
+// pruning from the front runs the risk of making a-replica-... and b-replica-... collide
+// so this removes characters from the middle of the string
+func NewPvcName(originalName string) string {
+	candidate := originalName + nameSuffix
+	if len(candidate) <= 253 {
+		return candidate
+	}
+
+	// remove characters from the middle of the string to reduce the total length to 63 characters
```
Comment (laverya) [inline pkg/k8sutil/truncate.go:17]:
```suggestion
	// remove characters from the middle of the string to reduce the total length to 253 characters
```

**Review by laverya — APPROVED** [review]

**Review by ricardomaraschini — APPROVED** [review]

**Comment by laverya** [comment]
Thank you for the PR @clementnuss !

**Pull request merged**